### PR TITLE
Add INT16,INT32 datatype

### PR DESCRIPTION
**Pull request by acquamarin**
# INT32(4 bytes),INT16(2 bytes)
## 1. Grammar rule for defining INT32/INT16 column:
To create a column with INT32/INT16 datatype, simply declare the column as: `COLUMN_NAME INT32/INT16`
## 3. Implements `to_int32()`, `to_int64()` function.

### CODE REVIEW

File: src/processor/operator/aggregate/aggregate_hash_table.cpp [hunk 1]
```diff
@@ -676,36 +676,31 @@ bool AggregateHashTable::compareEntryWithKeys(const uint8_t* keyValue, const uin
 
 compare_function_t AggregateHashTable::getCompareEntryWithKeysFunc(DataTypeID typeId) {
     switch (typeId) {
-    case INTERNAL_ID: {
+    case INTERNAL_ID:
```
Comment (andyfengHKU) [inline src/processor/operator/aggregate/aggregate_hash_table.cpp:679]: I think our code style is to have {} even each case is a one line return statement. So revert this change.

File: src/processor/operator/order_by/order_by_key_encoder.cpp [hunk 2]
```diff
@@ -210,6 +210,12 @@ encode_function_t OrderByKeyEncoder::getEncodingFunction(DataTypeID typeId) {
     case INT64: {
         return encodeTemplate<int64_t>;
     }
+    case INT32: {
+        return encodeTemplate<int32_t>;
+    }
+    case INT16: {
+        return encodeTemplate<int16_t>;
+    }
```
Comment (andyfengHKU) [inline src/processor/operator/order_by/order_by_key_encoder.cpp:218]: Put FLOAT under DOUBLE

File: src/include/function/hash/hash_operations.h [hunk 3]
```diff
@@ -69,6 +69,16 @@ inline void Hash::operation(const int64_t& key, common::hash_t& result) {
     result = murmurhash64(key);
 }
 
+template<>
+inline void Hash::operation(const int32_t& key, common::hash_t& result) {
+    result = murmurhash64(key);
+}
+
+template<>
+inline void Hash::operation(const int16_t& key, common::hash_t& result) {
```
Comment (andyfengHKU) [inline src/include/function/hash/hash_operations.h:78]: put float under double

File: src/function/built_in_vector_operations.cpp [hunk 4]
```diff
@@ -114,6 +118,10 @@ uint32_t BuiltInVectorOperations::getCastCost(
 
 uint32_t BuiltInVectorOperations::getTargetTypeCost(common::DataTypeID typeID) {
     switch (typeID) {
+    case common::INT32:
```
Comment (andyfengHKU) [inline src/function/built_in_vector_operations.cpp:121]: Add bracket. Also, are we copying these magic numbers from duckdb? I think we need to add some tests or issue about

INT16 + INT32 will bind to function INT32 + INT32.

File: src/function/vector_cast_operations.cpp [hunk 5]
```diff
@@ -42,9 +72,14 @@ scalar_exec_func VectorCastOperations::bindExecFunc(
 
 std::string VectorCastOperations::bindCastFunctionName(common::DataTypeID targetTypeID) {
     switch (targetTypeID) {
-    case common::DOUBLE: {
+    case common::INT64:
+        return CAST_TO_INT64_FUNC_NAME;
```
Comment (andyfengHKU) [inline src/function/vector_cast_operations.cpp:76]: {}

File: src/function/vector_cast_operations.cpp [hunk 6]
```diff
@@ -8,9 +8,39 @@ using namespace kuzu::common;
 namespace kuzu {
 namespace function {
 
-scalar_exec_func VectorCastOperations::bindExecFunc(
+scalar_exec_func VectorCastOperations::bindImplicitCastFunc(
     common::DataTypeID sourceTypeID, common::DataTypeID targetTypeID) {
     switch (sourceTypeID) {
+    case common::INT16: {
+        switch (targetTypeID) {
+        case common::INT32:
```
Comment (andyfengHKU) [inline src/function/vector_cast_operations.cpp:16]: Now you might want to warp the second switch case as its own function.

So you will have
```
castINT16(targetTypeID)
castINTXX()
```

**Review by andyfengHKU — APPROVED** [review]
A few other things.

Open another PR to make sure user can declare type INT which will bind to INT64 (or INT32 you can check what other systems pick) internally.

remove `dataset/copy-fault-tests/invalid-number/copy.cypher`

File: src/common/types/types.cpp [hunk 7]
```diff
@@ -81,6 +83,8 @@ DataType& DataType::operator=(const DataType& other) {
     case INTERNAL_ID:
     case BOOL:
     case INT64:
+    case INT32:
+    case INT16:
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/types/types.cpp:87]: done

File: test/test_files/tinysnb/order_by/single_label.test [hunk 8]
```diff
@@ -11,6 +11,22 @@
 45
 83
 
+-NAME OrderByInt32Test
+-QUERY MATCH (m:movies) RETURN m.length ORDER BY m.length
+-PARALLELISM 2
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply test/test_files/tinysnb/order_by/single_label.test:16]: removed

File: src/processor/operator/aggregate/aggregate_hash_table.cpp (hunk 1 above)
Comment (acquamarin) [reply src/processor/operator/aggregate/aggregate_hash_table.cpp:679]: reverted

File: src/processor/operator/order_by/order_by_key_encoder.cpp (hunk 2 above)
Comment (acquamarin) [reply src/processor/operator/order_by/order_by_key_encoder.cpp:218]: done

File: src/include/function/hash/hash_operations.h (hunk 3 above)
Comment (acquamarin) [reply src/include/function/hash/hash_operations.h:78]: done

File: src/function/vector_cast_operations.cpp (hunk 5 above)
Comment (acquamarin) [reply src/function/vector_cast_operations.cpp:76]: done

File: src/function/vector_cast_operations.cpp (hunk 6 above)
Comment (acquamarin) [reply src/function/vector_cast_operations.cpp:16]: done

File: src/include/common/type_utils.h [hunk 9]
```diff
@@ -45,42 +48,26 @@ class TypeUtils {
...
     template<typename T>
-    static T convertFloatingPointNumber(const char* data, DataTypeID typeID) {
-        char* eptr;
-        errno = 0;
-        auto retVal = convertStringToFloatingPointNumber<T>(data, &eptr);
-        throwConversionExceptionIfNoOrNotEveryCharacterIsConsumed(data, eptr, typeID);
-        if ((HUGE_VAL == retVal || -HUGE_VAL == retVal) && errno == ERANGE) {
-            throwConversionExceptionOutOfRange(data, typeID);
+    static T convertStringToNumber(const char* data) {
+        std::istringstream iss{data};
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/common/type_utils.h:52]: yes, this is the c++ way to convert string to numbers.
template functions can't be moved to cpp

[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/common/type_utils.h:52]: ditto

File: src/function/built_in_vector_operations.cpp (hunk 4 above)
Comment (acquamarin) [reply src/function/built_in_vector_operations.cpp:121]: added bracket
